### PR TITLE
Feat/kpcovr

### PR DESCRIPTION
**Pull request by rosecers**
Initial implementation of kpcovr. @bhelfrecht I'm going to assign this PR directly to you so that you can rifle through and make sure everything implemented is in line with what we've developed. There are some edits to the pcovr frameworks that could go into a separate PR, but are necessary for kpcovr to function.

### CODE REVIEW

File: skcosmo/pcovr/kpcovr.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+                )
+            else:
+                W = np.linalg.lstsq(K, Yhat, rcond=self.regularization_)[0]
+
+        if Yhat is None:
+            Yhat = K @ W
+
+        self._fit(K, Yhat, W)
+
+        self.ptk_ = self.pt__ @ K
```
Comment (bhelfrecht) [inline skcosmo/pcovr/kpcovr.py:278]:
```suggestion
        self.ptk_ = np.linalg.lstsq(T, K, rcond=self.regularization_)[0]
```

File: skcosmo/pcovr/kpcovr.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        self.pty_ = self.pt__ @ Y
```
Comment (bhelfrecht) [inline skcosmo/pcovr/kpcovr.py:279]:
```suggestion
        self.pty_ = np.linalg.lstsq(T, Y, rcond=self.regularization_)[0]
```

File: skcosmo/pcovr/kpcovr.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+
+        if self.n_components is None:
+            self.n_components = min(X.shape)
+
+        self._centerer = KernelFlexibleCenterer()
+        K = self._get_kernel(X)
+
+        if W is None:
+            if Yhat is None:
+                W = (np.linalg.pinv(K, rcond=self.regularization_) @ Y).reshape(
```
Comment (bhelfrecht) [inline skcosmo/pcovr/kpcovr.py:267]:
```suggestion
                W = (np.linalg.lstsq(K, Y, rcond=self.regularization_)[0]).reshape(
```
This is probably a good spot for a `np.linalg.lstsq` regardless of what we decide to do about the `p` matrices. On a similar note, is there a reason why this is a `pinv` and the one in the `else` below is a `lstsq`?

File: skcosmo/pcovr/kpcovr.py [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+            and n_components is the number of components.
+
+        Returns
+        -------
+        X_original array-like, shape (n_samples, n_features)
+        """
+
+        return T @ self.ptx_
+
+    def score(self, X, Y):
```
Comment (bhelfrecht) [inline skcosmo/pcovr/kpcovr.py:353]: Do we want the "real" KPCovR loss in addition to (or in place of) the one here?

File: skcosmo/pcovr/kpcovr.py [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+
+        """
+
+        check_is_fitted(self, ["pkt_", "X_fit_"])
+
+        X = check_array(X)
+        K = self._get_kernel(X, self.X_fit_)
+        return K @ self.pkt_
+
+    def inverse_transform(self, T):
```
Comment (bhelfrecht) [inline skcosmo/pcovr/kpcovr.py:326]: Do we also want an inverse transform to the kernel using `self.ptk_`?

File: tests/test_kpcovr.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    def test_kernel_types(self):
+        """
+        This test checks that KPCovR can handle all kernels passable to
+        sklearn kernel classes
+        """
+        kernel_params = {
+            "poly": {"degree": 2},
+            "rbf": {"gamma": 3.0},
+            "sigmoid": {"gamma": 3.0, "coef0": 0.5},
+        }
```
Comment (bhelfrecht) [inline tests/test_kpcovr.py:119]: Should we also test for precomputed kernels and kernels constructed from callable functions? (`KernelRidge` also accepts these as kernel options, for example -- from a quick glance the KPCovR class might be able to handle it as well)

File: tests/test_kpcovr.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+        # computing projection and predicton loss with PCovR
+        ref_model = PCovR(**hypers)
+        ref_model.fit(self.X, self.Y, Yhat=Yhat)
+        t_ref = ref_model.transform(self.X)
+        _, ly_ref = ref_model.score(self.X, self.Y)
+
+        # computing the loss in K similar to that in kpcovr
+        K = model._get_kernel(self.X, model.X_fit_)
+        k_ref = t_ref @ np.linalg.pinv(t_ref.T @ t_ref, rcond=1e-8) @ t_ref.T @ K
```
Comment (bhelfrecht) [inline tests/test_kpcovr.py:160]:
```suggestion
        k_ref = t_ref @ np.linalg.lstsq(t_ref, K, rcond=1e-8)[0]
```

File: skcosmo/pcovr/kpcovr.py [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+        K = self._centerer.fit_transform(K)
+        K_tilde = pcovr_kernel(mixing=self.mixing_, X=K, Y=Yhat, kernel="precomputed")
+
+        v, U = eig_solver(
+            K_tilde, tol=self.regularization_, n_components=self.n_components
+        )
+
+        P = (self.mixing_ * np.eye(K.shape[0])) + (1.0 - self.mixing_) * (W @ Yhat.T)
+
+        v_inv = np.linalg.pinv(np.diagflat(v))
```
[the remark this answers is not in the document]
Comment (bhelfrecht) [reply skcosmo/pcovr/kpcovr.py:211]: As a quick addendum, `np.linalg.pinv` indeed appears to be about 100x-1000x slower than the plain inversion (for 100-1000 eigenvalues)

File: skcosmo/pcovr/kpcovr.py (hunk 4 above)
Comment (bhelfrecht) [reply skcosmo/pcovr/kpcovr.py:353]: One other comment since I've recently been looking into sklearn's scoring infrastructure. In sklearn, the `score` method for each estimator should only return a _single  value_, and this is required for compatibility with the existing pipeline and cross validation infrastructure. I suggest that we separate the (K)PCovR losses into separate methods, e.g., `kpcovr.lkpca` and `kpca.lkrr`, and then essentially make `kpcovr.score` a wrapper around these that returns the sum (or mean) of the two. Things would get a little more complex if we add additional arguments to `kpcovr.score` (e.g., throwing the Gram loss in the mix or a weighted mean of the losses), but I think it would still be usable in sklearn pipelines and cross validation.

File: skcosmo/pcovr/kpcovr.py [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+    def _fit(self, K, Yhat, W):
+        """
+        Fit the model with the computed kernel and approximated properties.
+        """
+
+        K = self._centerer.fit_transform(K)
+        K_tilde = pcovr_kernel(mixing=self.mixing_, X=K, Y=Yhat, kernel="precomputed")
+
+        v, U = eig_solver(
+            K_tilde, tol=self.regularization_, n_components=self.n_components
```
[the remark this answers is not in the document]
Comment (rosecers) [reply skcosmo/pcovr/kpcovr.py:206]: Good catch -- this should map to the `self.tol` parameter, not `self.regularization_`

File: skcosmo/pcovr/kpcovr.py (hunk 3 above)
Comment (rosecers) [reply skcosmo/pcovr/kpcovr.py:267]: Only reason is me breaking out of the habit of pinv...

File: skcosmo/pcovr/kpcovr.py (hunk 4 above)
Comment (rosecers) [reply skcosmo/pcovr/kpcovr.py:353]: That's a solid point -- although I'm split on whether to use the _true_ KPCovR loss. On one hand, it _is_ the loss being optimized, but on the other hand, it is not necessarily the loss that people are interested in. What are your thoughts?

File: skcosmo/pcovr/kpcovr.py (hunk 5 above)
Comment (rosecers) [reply skcosmo/pcovr/kpcovr.py:326]: sklearn KPCA does "inverse transform" back to X, despite the caveats, so I figured we would do the same here.

File: skcosmo/pcovr/kpcovr.py [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+        )
+
+        P = (self.mixing_ * np.eye(K.shape[0])) + (1.0 - self.mixing_) * (W @ Yhat.T)
+
+        v_inv = np.linalg.pinv(np.diagflat(v))
+
+        self.pkt_ = P @ U @ np.sqrt(v_inv)
+
+        T = K @ self.pkt_
```
[the remark this answers is not in the document]
Comment (rosecers) [reply skcosmo/pcovr/kpcovr.py:213]: What do you think of my fix? L215

**Review by bhelfrecht — APPROVED** [review]

File: skcosmo/pcovr/kpcovr.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (bhelfrecht) [reply skcosmo/pcovr/kpcovr.py]: Excellent
